0001-dinerojs-v2.md

```diff
@@ -0,0 +1,188 @@
+
+```js
+import { createDinero } from "@dinero.js/helpers";
+import Big from "some-bigint-library";
+
+const Dinero = createDinero({
+  calculator: {
+    add: Big.add,
+    subtract: Big.subtract
+    // ...
+  }
+});
+```
+
+A `Calculator` type should be provided to ease the creation with TypeScript:
+
+```ts
+const BigIntCalculator: Calculator = {
+  add: Big.add,
+  subtract: Big.subtract
+  // ...
+};
+```
+
+#### Currency service
+
+Currently, users must specify the `precision` of their `Dinero` objects, which is set to `2` by default. When a user creates objects with a currency which exponent is different from `2` (e.g., Iraqi dinars, Japanese yens, etc.), they must specify the `precision` by hand. One way to alleviate it is by setting a `defaultPrecision`, but this only works when one is using a single currency throughout their project.
+
+The currency service lets the user pass a `Currency` object instead of a plain string to the `currency` property when creating `Dinero` objects. `Currency` objects embark the currency's exponent, allowing the `Dinero` object to set the `precision` property automatically.
+
+```js
+import Dinero from "dinero.js";
+import { IQD } from "@dinero.js/currencies";
+
+// `precision` is set to `2` (default)
+const d1 = Dinero({ amount: 1000, currency: "IQD" });
+
+// `precision` is automatically set to `3`
+const d2 = Dinero({ amount: 1000, currency: IQD });
+```
+
+`Currency` objects could also embark the currency's symbol, allowing the library to fully own the currency formatting within the [`toFormat`](https://sarahdayan.github.io/dinero.js/module-Dinero.html#~toFormat) method and liberate itself from its dependency on the I18n API.
+
+```js
+import Dinero from "dinero.js";
+import { EUR } from "@dinero.js/currencies";
+
+const d1 = Dinero({ amount: 100, currency: "EUR" });
+const d2 = Dinero({ amount: 100, currency: EUR });
+
+d1.toFormat("$0,0.00"); // EUR100.00
+d2.toFormat("$0,0.00"); // €100.00
+```
+
+`Dinero` objects created with a string `currency` instead of a `Currency` object don't have any information on the currency symbol. Consequently, they would output the currency code in full instead of the symbol.
+
+The I18n API could still be used on Dinero objects as with v1, by chaining it to a `toRoundedUnit` method call: `Dinero().toRoundedUnit(digits, roundingMode).toLocaleString(locale, options)`.
+
+Users would also be able to create their own, custom currencies, breaking out of the ISO 4217 limit.
+
+```js
+const FOO = {
+  name: "foo",
+  symbol: "F",
+  code: "FOO",
+  exponent: 4
+};
+```
+
+TypeScript would further facilitate it, as they would have access to a `Currency` type.
+
+```ts
+const FOO: Currency = {
+  name: "foo",
+  symbol: "F",
+  code: "FOO",
+  exponent: 4
+};
+```
+
+#### Helpers
```

Comment:
My biggest beef with v1 (which is otherwise an awesome library) is the fact that you don't use constuctors. I dislike verifying method parameters using duck typing. So here is some concrete criticism:

1. Helper factories for returning instances are fine but they should return an `instanceof Dinero`.
2. `DineroUSD`, `DineroEUR` etc feels odd. I would expect `Dinero.USD`, `Dinero.EUR` and so on. And again they would all be instance of type `Dinero`.

Reply:
Hey @cowwoc, thanks for this thorough message! I have a few questions:

1. When and why do you need to verify whether an object is `instanceof Dinero`? It seems to me that using TypeScript (which Dinero v2 will bring natively, but can also be done with the current third-party typings) is a much better way to do structural typing. In the case of runtime type checking, I believe the dynamic nature of prototype delegates makes `instanceof` quite unreliable. Duck typing seems much more suitable. Could you share your use case(s) with me? I'd be interested in debating this matter over an actual piece of code 🙂
2. This is rather odd that your IDE isn't auto-completing, given that a Dinero instance is a plain object. What IDE is it? Things work well in VS Code (c.f. the following screenshot from CodeSandbox). I haven't checked with WebStorm, but any IDE or editor with a JavaScript extension should provide auto-completion when trying to access object properties. This isn't specific to Dinero.
![Capture d’écran 2019-09-28 à 16 08 53](https://user-images.githubusercontent.com/5370675/65817668-7b700500-e20a-11e9-95ce-fc02f6e95f26.png)
3. Adding currency helper onto the main `Dinero` object would be highly problematic, because it would weight it down with hundreds of currency definitions. To give you an idea, there are currently 165 ISO 4217 currencies and 9 crypto currencies in Dinero v2! You likely need a couple in a regular application, so it makes no sense for you to carry around a bunch of superfluous factories that you can't even tree-shake. I intend to keep currencies apart from the core, precisely because you want to only use what you need; same goes for helpers. Does that make sense?

Reply:
duck typing is interesting, but more challenging to deal with in certain cases, I'd personally rather these were classes. I tried to implement/or extend Dinero such that I had an interface, Payment, that isa Money, but that went poorly for me in 1.0.

Reply:
also, not being an expert, it might be useful to look at the upcoming? Money Api for java https://www.baeldung.com/java-money-and-currency

Reply:
@xenoterracide I definitely understand that coming from a Java background, classes feel more natural. Yet since this is JavaScript, the choices I'm making are different from the kind of choices you make in a 100% object-oriented language. For instance, many JavaScript developers today are embracing a functional approach, where object-oriented patterns don't make sense at all.

Can you share with me specific use cases where `instanceof` would have helped you and where duck typing fell short? And how you wanted to extend Dinero but couldn't?

Reply:
I come from java, before that perl (and dabbled in a dozen or so total, including PHP), both support functional programming, currently working with typescript. I've never seen FP that didn't have "objects" according to what I believe the definition of an object is.

I'm not sure I need `instanceof` what I'm doing is this, currently I'm wrapping/proxying to dinero, due to its lack of support for custom currencies, but I'd just as soon replace money with dinero.

```
export default interface Quantity {
  amount: number;
}
export default interface Money extends Quantity {
  readonly currency: string;
}
export default interface Payment<T extends PaymentMethod> extends Money {
  readonly paymentMethod: T;
}
```

but my initial attempt was to extend `Dinero.Dinero` instead of `Money`, the problem that came in was that then I was expected to implement all the things it seemed when implementing `Payment`. so instead, for the time being I've created a class that proxies to dinero (though I've thought of just using `Proxy`)

 my model is stolen shamelessly from https://amzn.to/2ZVTXRO, which is a good read for implementing currency as well.

Reply:
Dinero v2 will be fully written in TypeScript, and therefore come with a `DineroFactory` interface. Would this solve your issue?

Reply:
I mean, the `Payment` class isn't a factory, it's a `Money` so I'm not sure... maybe? I'm using the @types/dinero.js now, the problem isn't really a lack of an interface it's the lack of ability to extend (not necessarily referring to the keyword) the actual behavior without delegation. Maybe there's a better way in typescript now, but I'm not aware of it.

Reply:
note: if there's an interface, that's not really duck typing

but further thoughts on my current problem, currently I could make my `interface Money extend DineroOptions` and then instantiate Dinero when I need to do math, with 2.0 I could probably make it so that my `Payment` type `has`a `Money` because now we can represent our "bucks" as a custom type of Money. still wondering though if there might be a way of extending the new dinero money such that I don't have to "reimplement" everything... maybe a proxy helper... ... or does typescript support mixins/traits?

Reply:
@xenoterracide In Typescript interfaces just define a shape. So as long as you pass something else in that matches that shape, Typescript considers it a match and is happy, regardless of if you declared usage of that particular interface in the original function or not.

I think that's what @sarahdayan means by duck typing here. You could build a custom class and start passing instances into Dinero methods without any issue.

Reply:
@thekevinBrown, sure, but I'm trying to do something a little cleaner than this, which I think is a little cleaner than writing delegating classes or proxies

```
  const auth = await processor.authorize({
    ...Dinero({ amount: 350, currency: 'USD' }),
    paymentMethod: {
      token: 'pm_card_visa',
    },
  });
```

but if I could either extend a class to simply add 1 field, or perhaps otherwise be able to write...

```
  const auth = await processor.authorize({
    amount: 350,
    currency: 'USD',
    paymentMethod: {
      token: 'pm_card_visa',
    },
  });
```

Reply:
@xenoterracide Sorry, if I'm being silly, but I'm not following what your use case is. You don't need to write a delegating class or proxy for what you're doing. Why not just do:

```
const auth = await processor.authorize({
    amount: Dinero({ amount: 350, currency: 'USD' }),
    paymentMethod: {
      token: 'pm_card_visa',
    },
  });
```
And have your processor accept a Dinero instance?

Reply:
well right now, because Dinero doesn't support custom currencies, and we have one... I mean, in the future that is definitely a possibility that I've considered, maybe it's the right one... I'm not sure, the model is prescriptive that a Payment extends Money, authorize takes a Payment, but that might be a silly prescription.

Reply:
although it's a little weird to have `payment.amount.getAmount()`